### ISSUE TITLE
ConvNd ops in channel last layout (N,L,C) / (N,H,W,C) / (N,D,H,W,C)

### ISSUE THREAD

**Issue by mauriceweiler** [pytorch/pytorch#157663]
### 🚀 The feature, motivation and pitch

Torch's nn.ConvNd modules (and other grid ops) operate in the channel first layout `(N,C,L)` / `(N,C,H,W)` / `(N,C,D,H,W)` while other operations assume channels in the last axis. This becomes problematic when combining such operations. For instance, one may want to build a vision transformer which alternates Conv2d and attention operations, which requires expensive transpositions/memcopy back and forth between `(N,C,H,W)` and `(N, H*W, C)`.

### Proposition:
Provide convolution operations (and other ops) that operate on tensors in the channel last layout  `(N,L,C)` / `(N,H,W,C)` / `(N,D,H,W,C)`. This could be done via an additional flag in `__init__`
``` python
class ConvNd(nn.Module):
    def __init__(..., channels_last: bool = False):
        ...
```
or by implementing a new module
``` python
class ConvNdChannelsLast(nn.Module):
    ...
```
Both options would be fully backward compatible with the current interface.
This should only be a matter of dispatching new kernels, which do already exist on multiple devices:
https://docs.nvidia.com/deeplearning/performance/dl-performance-convolutional/index.html#tensor-layout
https://docs.pytorch.org/tutorials/recipes/intel_extension_for_pytorch.html#features

This feature has been requested multiple times but is still not present.
https://github.com/pytorch/pytorch/issues/50962
https://github.com/pytorch/pytorch/issues/1220
https://discuss.pytorch.org/t/dataset-with-last-dimension-as-channel/100639
https://github.com/pytorch/pytorch/issues/74935

Besides consistency with other pytorch operations, convolutions are significantly more performant in the channels last layout. This is especially true if Tensor Cores are used (`(N,C,H,W)` kernels using tensor cores are internally permuting data to `(N,H,W,C)`). 
https://docs.pytorch.org/tutorials/intermediate/memory_format_tutorial.html#performance-gains
https://docs.nvidia.com/deeplearning/performance/dl-performance-convolutional/index.html#tensor-layout

Is there a good reason why this feature is / will not be implemented? I didn't see any convincing answer in all the discussions linked above.

### Alternatives

Torch offers an experimental `torch.channels_last` memory format which is closely related ot the requested feature.
https://docs.pytorch.org/tutorials/intermediate/memory_format_tutorial.html

The tutorial shows how to transform channel first tensors to what is essentially *contiguous* channel last `(N,H,W,C)` with stride `(HWC, WC, C, 1)`, but is interpreted as a shape `(N,C,H,W)` tensor with non-contiguous stride `(HWC, 1, WC, C)`.
``` python
x = torch.randn(N,C,H,W)                    # shape (N,C,H,W) stride (HWC, HW, W, 1)
x = x.to(memory_format=torch.channels_last) # shape (N,C,H,W) stride (HWC, 1, WC, C), corresponding to contiguous (N,H,W,C) stride (HWC, WC, C, 1)
```
Using such input with a ConvNd instance in the same layout (`conv_op.to(memory_format=torch.channels_last)`) dispatches to the desired kernel in the channel last layout.

The issue with this approach is that 1) it is rather awkward and confusing and 2) starts from the channels first layout `(N,C,H,W)` and then applies a costly permutation to the channels last layout.
https://discuss.pytorch.org/t/dataset-with-last-dimension-as-channel/100639/15

I found the following workaround which starts with a `(N,H,W,C)` tensor and reinterprets it as `torch.channels_last`:
```python
x = torch.randn(N,H,W,C)                          # shape (N,H,W,C) stride (HWC, WC, C, 1)
x = torch.as_strided(input  = x,                  # shape (N,C,H,W) stride (HWC, 1, WC, C)
                     size   = (N,C,H,W),
                     stride = (H*W*C, 1, W*C, C))
assert x.is_contiguous(memory_format=torch.channels_last) # True
```
This allows to apply a convolution to `(N,H,W,C)` tensors *without additional permutation/memcopy*.
After the convolution, the output can similarly be reinterpreted as a *contiguous* `(N,H',W',C_out)` tensor, as desired.
I tested results and all gradients to agree with the channel first implementation.

The issue with this approach is that it does not work for all combinations of devices, dimensions and dtypes:
- on CPU and in dim=2, most dtypes work, but for dim=3 only `float32` and `bfloat16` seem to work.
- for CUDA devices, only `float32` seems to work.
- for dim=1, there is no `channels_last_1d` format. Related issue: https://github.com/pytorch/pytorch/issues/74935
 
Besides this, the `channels_last` approach is a syntax nightmare compared to the clean approach proposed above.

### Additional context

According to this issue, there is yet another layout `(L,N,C)` (instead of `(N,C,L)` or `(N,L,C)`) implemented:
https://github.com/pytorch/pytorch/issues/34306

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @jamesr66a

**Comment by mauriceweiler** [comment]
Any comments or opinions regarding this feature request? Is there any good reason to not implement ConvNd modules with channels-last layout?

**Comment by mauriceweiler** [comment]
Any comments or opinions regarding this feature request? Is there any good reason to not implement ConvNd modules with channels-last layout?

[sorry, closed accidentally]
